content/post/2020-03-05-user-preferences/index.Rmd

```diff
@@ -0,0 +1,254 @@
+
+* or write that same line in their [.Rprofile that's loaded before every session](https://rstats.wtf/r-startup.html), which is more persistent. 
+
+Users can use a project-level or more global user-level .Rprofile. 
+The use of a project-level .Rprofile overrides the user-level .Rprofile unless the project-level .Rprofile contains the following lines [as mentioned in the `blogdown` book](https://bookdown.org/yihui/blogdown/global-options.html):
+
+```r
+# in .Rprofile of the project
+if (file.exists('~/.Rprofile')) {
+  base::sys.source('~/.Rprofile', envir = environment())
+}
+# then set project options
+```
+
+For more startup tweaks, the user could adopt [the `startup` package](https://cran.r-project.org/web/packages/startup/index.html).
+
+As a package developer in your code you can retrieve options by using `getOption()` whose second argument is a fallback for when the option hasn't been set by the user.
+Note that an option can be any R object.
+
+```{r options}
+options(blabla.foo = TRUE)
+if (getOption("blabla.foo", FALSE)) {
+  message("foo!")
+}
+
+options(blabla.bar = mean)
+getOption("blabla.bar")(c(1:7))
+```
+
+The use of options in the .Rprofile startup file is great for workflow packages like `usethis`, `blogdown`, etc., but shouldn't be used for, say, arguments influencing the results of a statistical function.
+
+### Using environment variables
+
+Environment variables, found via `Sys.getenv()` rather than `getOption()`, are appropriate for storing secrets (like [`GITHUB_PAT` for the `gh` package](https://github.com/r-lib/gh#tokens)) or the path to secrets on disk (like [`TWITTER_PAT` for `rtweet`](https://rtweet.info/articles/auth.html)), or not secrets (e.g. [the browser to use for `chromote`](https://github.com/rstudio/chromote#specifying-which-browser-to-use)).
+
+Similar to using `options()` in the console or at the top of a script the user could use `Sys.setenv()`.
+Obviously, secrets should not be written at the top of a script that's public.
+To make environment variables persistent they need to be stored in a startup file, .Renviron.
+.Renviron does not contain R code like .Rprofile, but rather key-value pairs that are only called via `Sys.getenv()`.
+
+As a package developer, you probably want to at least document how to set persistent variables or provide a link to such documentation; and you could even provide helper functions like [what `rtweet` does](https://rtweet.info/reference/create_token.html).
+
+### Using credential stores for secrets
+
+Although say API keys can be stored in `.Renviron`, they could also be stored in a standard location depending on the operating system.
+The [`keyring` package](https://github.com/r-lib/keyring#keyring) allows to interact with such credential stores.
+You could either take it on as a dependency or recommend the user of your package to use `keyring` and to add a line like
+
+```r
+Sys.setenv(SUPERSECRETKEY = keyring::key_get("myservice"))
+```
```

Comment:
Even better, gh can now get the token directly from the keyring, so no env var is needed. https://github.com/r-lib/gh/blob/bfff4473c2876e4fbdb48d952c86e804d75174ad/R/gh_token.R#L34

So we can hopefully get rid of secrets in env vars. But this might be slightly off topic here.